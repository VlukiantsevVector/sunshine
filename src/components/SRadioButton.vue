<template>
  <label :class="classes(modifiers)">
    <input
      type="radio"
      :name="name"
      :checked="checked"
      :disabled="inactive"
      @change="onChange"
      @focus="hasFocus = true"
      @blur="hasFocus = false"
    >

    <span :class="classes('box')">
      <transition name="fade">
        <svg
          v-if="checked"
          :class="classes('selected')"
          width="16"
          height="16"
        >
          <path
            fill="#1C1C1C"
            d="M8,11 C6.34314575,11 5,9.65685425 5,8 C5,6.34314575
              6.34314575,5 8,5 C9.65685425,5 11,6.34314575 11,8 C11,9.65685425
              9.65685425,11 8,11 Z"
          />
        </svg>
      </transition>
    </span>
    <slot />
  </label>
</template>

<script>
import Vue from 'vue';
import mixins from 'vue-typed-mixins';
import { ClassesMixin } from '../lib/utils';

export default mixins(ClassesMixin).extend({
  name: 's-radio-button',

  model: {
    prop: 'choice',
    event: 'input',
  },

  props: {
    value: String,
    choice: String,
    name: {
      type: String,
      default: '',
    },
    inactive: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      checked: this.choice === this.value,
      hasFocus: false,
    };
  },

  watch: {
    choice(val) {
      this.checked = val === this.value;
    },
  },

  computed: {
    modifiers() {
      return {
        focus: this.hasFocus,
        inactive: this.inactive,
      };
    },
  },

  methods: {
    onChange() {
      this.checked = true;
      this.$emit('input', this.value);
    },
  },
});
</script>
