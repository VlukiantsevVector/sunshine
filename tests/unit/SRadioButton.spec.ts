import { shallowMount } from '@vue/test-utils';
import SRadioButton from '@/components/SRadioButton.vue';

let wrapper : any;
beforeAll(() => {
  wrapper = shallowMount(SRadioButton, {
    propsData: {
      value: 'test',
      checked: false,
    },
  });
});

describe('SRadioButton', () => {
  it('render unchecked', () => {
    expect(wrapper.find('.s-radio-button__selected').exists()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('render checked dot when selected', () => {
    wrapper.find('input').setChecked();
    expect(wrapper.find('.s-radio-button__selected').exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('slot content is rendered', () => {
    const _wrapper = shallowMount(SRadioButton, {
      slots: {
        default: '<h1>Test</h1>',
      },
    });

    expect(_wrapper.html()).toMatchSnapshot();
  });

  it('emits value when clicked', () => {
    wrapper.find('input').setChecked();
    expect(wrapper.emitted('input')[0][0]).toBe('test');
  });

  it('render checked', () => {
    const _wrapper = shallowMount(SRadioButton, {
      propsData: {
        value: 'test',
        choice: 'test',
      },
    });
    expect(_wrapper.find('.s-radio-button__selected').exists()).toBeTruthy();
    expect(_wrapper.html()).toMatchSnapshot();
  });
});
