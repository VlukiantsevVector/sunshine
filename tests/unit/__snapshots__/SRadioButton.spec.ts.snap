// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SRadioButton render checked 1`] = `
<label class="s-radio-button"><input type="radio" name=""> <span class="s-radio-button__box"><svg width="16" height="16" class="s-radio-button__selected" name="fade"><path fill="#1C1C1C" d="M8,11 C6.34314575,11 5,9.65685425 5,8 C5,6.34314575
            6.34314575,5 8,5 C9.65685425,5 11,6.34314575 11,8 C11,9.65685425
            9.65685425,11 8,11 Z"></path></svg></span> </label>
`;

exports[`SRadioButton render checked dot when selected 1`] = `
<label class="s-radio-button"><input type="radio" name=""> <span class="s-radio-button__box"><svg width="16" height="16" class="s-radio-button__selected" name="fade"><path fill="#1C1C1C" d="M8,11 C6.34314575,11 5,9.65685425 5,8 C5,6.34314575
            6.34314575,5 8,5 C9.65685425,5 11,6.34314575 11,8 C11,9.65685425
            9.65685425,11 8,11 Z"></path></svg></span> </label>
`;

exports[`SRadioButton render unchecked 1`] = `<label class="s-radio-button"><input type="radio" name=""> <span class="s-radio-button__box"><!----></span> </label>`;

exports[`SRadioButton slot content is rendered 1`] = `
<label class="s-radio-button"><input type="radio" name=""> <span class="s-radio-button__box"><svg width="16" height="16" class="s-radio-button__selected" name="fade"><path fill="#1C1C1C" d="M8,11 C6.34314575,11 5,9.65685425 5,8 C5,6.34314575
            6.34314575,5 8,5 C9.65685425,5 11,6.34314575 11,8 C11,9.65685425
            9.65685425,11 8,11 Z"></path></svg></span>
  <h1>Test</h1></label>
`;
